File: openc3-json-tlm-server/python/json_telemetry_server_interface.py
# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import json
import queue
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from threading import Thread

from openc3.interfaces.http_server_interface import HttpServerInterface
from openc3.system.system import System


class _JsonHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):  # noqa: A002
        pass  # suppress access log noise

    def do_POST(self):  # noqa: N802
        self._handle()

    def do_GET(self):  # noqa: N802
        self.send_error_response(405, {"Allow": "POST"})

    def do_PUT(self):  # noqa: N802
        self.send_error_response(405, {"Allow": "POST"})

    def do_DELETE(self):  # noqa: N802
        self.send_error_response(405, {"Allow": "POST"})

    def do_PATCH(self):  # noqa: N802
        self.send_error_response(405, {"Allow": "POST"})

    def send_error_response(self, status, extra_headers=None):
        self.send_response(status)
        if extra_headers:
            for k, v in extra_headers.items():
                self.send_header(k, v)
        self.send_header("Content-Length", "0")
        self.end_headers()

    def _handle(self):
        api_key = self.server.api_key
        if api_key is not None:
            auth = self.headers.get("Authorization", "")
            token = auth[7:] if auth.startswith("Bearer ") else None
            if token != api_key:
                self.send_response(401)
                self.send_header("WWW-Authenticate", "Bearer")
                self.send_header("Content-Length", "0")
                self.end_headers()
                return

        parts = [p for p in self.path.split("?")[0].split("/") if p]
        if len(parts) != 2:
            self.send_error_response(404)
            return

        target_name, packet_name = parts[0].upper(), parts[1].upper()
        try:
            System.telemetry.packet(target_name, packet_name)
        except Exception:
            self.send_error_response(404)
            return

        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length) if length else b""
        try:
            json.loads(body)
        except (json.JSONDecodeError, ValueError):
            self.send_error_response(400)
            return

        extra = {
            "HTTP_REQUEST_TARGET_NAME": target_name,
            "HTTP_REQUEST_PACKET_NAME": packet_name,
        }
        self.server.request_queue.put((body, extra))

        self.send_response(200)
        self.send_header("Content-Length", "0")
        self.end_headers()


class JsonTelemetryServerInterface(HttpServerInterface):
    """HTTP server that accepts POST /<target>/<packet> with a flat JSON body
    and maps JSON keys to packet items by name (case-insensitive). No
    HTTP-specific fields required in packet definitions."""

    def __init__(self, port=8765):
        super().__init__(port)
        self._api_key = None

    def set_option(self, option_name, option_values):
        super().set_option(option_name, option_values)
        if option_name.upper() == "API_KEY":
            self._api_key = option_values[0]

    def connect(self):
        self.server = ThreadingHTTPServer((self.listen_address, self.port), _JsonHandler)
        self.server.request_queue = queue.Queue()
        self.server.api_key = self._api_key
        self.server_thread = Thread(target=self.server.serve_forever, daemon=True)
        self.server_thread.start()
        # Call Interface.connect() directly (skip HttpServerInterface.connect which
        # tries to build the path lookup and mount routes from command packets)
        from openc3.interfaces.interface import Interface

        Interface.connect(self)

    def convert_data_to_packet(self, data, extra=None):
        from openc3.packets.packet import Packet

        if not extra:
            return Packet(None, None, "BIG_ENDIAN", None, data)

        target_name = extra.get("HTTP_REQUEST_TARGET_NAME")
        packet_name = extra.get("HTTP_REQUEST_PACKET_NAME")

        if not (target_name and packet_name):
            return Packet(None, None, "BIG_ENDIAN", None, data)

        try:
            packet = System.telemetry.packet(target_name, packet_name)
            packet = packet.clone()
        except Exception:
            return Packet(None, None, "BIG_ENDIAN", None, data)

        try:
            json_data = json.loads(data)
        except (json.JSONDecodeError, ValueError):
            return packet

        self._map_json_to_packet(json_data, packet)
        return packet

    def _map_json_to_packet(self, json_data, packet):
        """Write json_data values into packet items by case-insensitive name match."""
        item_lookup = {name.upper(): name for name in packet.items}
        for key, value in json_data.items():
            item_name = item_lookup.get(key.upper())
            if item_name:
                packet.write(item_name, value)

File: openc3-json-tlm-server/lib/json_telemetry_server_interface.rb
# encoding: ascii-8bit

# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/interfaces/http_server_interface'
require 'openc3/interfaces/interface'
require 'json'
require 'webrick'

module OpenC3
  # Accepts POST /<target_name>/<packet_name> with a flat JSON body and maps
  # JSON keys to packet items by name (case-insensitive). No HTTP-specific
  # fields required in packet definitions.
  class JsonTelemetryServerInterface < HttpServerInterface
    def initialize(port = 8765)
      super(port)
      @api_key = nil
    end

    # Supported Options:
    # API_KEY       - Bearer token required in Authorization header
    # LISTEN_ADDRESS - inherited from HttpServerInterface
    def set_option(option_name, option_values)
      super(option_name, option_values)
      if option_name.upcase == 'API_KEY'
        @api_key = option_values[0]
      end
    end

    def connect
      @server = WEBrick::HTTPServer.new(
        :BindAddress => @listen_address,
        :Port => @port,
        :Logger => WEBrick::Log.new('/dev/null'),
        :AccessLog => []
      )
      @request_queue = Queue.new

      api_key = @api_key
      request_queue = @request_queue

      @server.mount_proc '/' do |req, res|
        handle_request(req, res, api_key, request_queue)
      end

      Interface.instance_method(:connect).bind(self).call

      Thread.new do
        @server.start
      end
    end

    def convert_data_to_packet(data, extra = nil)
      unless extra
        return Packet.new(nil, nil, :BIG_ENDIAN, nil, data.to_s)
      end

      target_name = extra['HTTP_REQUEST_TARGET_NAME']
      packet_name = extra['HTTP_REQUEST_PACKET_NAME']

      unless target_name && packet_name
        return Packet.new(nil, nil, :BIG_ENDIAN, nil, data.to_s)
      end

      begin
        packet = System.telemetry.packet(target_name, packet_name)
        packet = packet.clone
      rescue
        return Packet.new(nil, nil, :BIG_ENDIAN, nil, data.to_s)
      end

      begin
        json_data = JSON.parse(data.to_s)
        map_json_to_packet(json_data, packet)
      rescue JSON::ParserError
        # already rejected upstream; return packet with defaults
      end

      packet
    end

    private

    def map_json_to_packet(json_data, packet)
      item_lookup = {}
      packet.items.each do |name, item|
        item_lookup[item.key.upcase] = name
      end
      json_data.each do |key, value|
        item_name = item_lookup[key.upcase]
        packet.write(item_name, value) if item_name
      end
    end

    def handle_request(req, res, api_key, request_queue)
      unless req.request_method == 'POST'
        res.status = 405
        res['Allow'] = 'POST'
        res.body = ''
        return
      end

      if api_key
        auth = req['Authorization'] || ''
        token = auth.start_with?('Bearer ') ? auth[7..] : nil
        unless token == api_key
          res.status = 401
          res['WWW-Authenticate'] = 'Bearer'
          res.body = ''
          return
        end
      end

      parts = req.path.split('/').reject(&:empty?)
      unless parts.length == 2
        res.status = 404
        res.body = ''
        return
      end

      target_name = parts[0].upcase
      packet_name = parts[1].upcase

      begin
        System.telemetry.packet(target_name, packet_name)
      rescue
        res.status = 404
        res.body = ''
        return
      end

      body = req.body.to_s
      begin
        JSON.parse(body)
      rescue JSON::ParserError
        res.status = 400
        res.body = ''
        return
      end

      extra = {
        'HTTP_REQUEST_TARGET_NAME' => target_name,
        'HTTP_REQUEST_PACKET_NAME' => packet_name,
      }
      request_queue << [body.dup, extra]

      res.status = 200
      res.body = ''
    end
  end
end
